Peloton Garbage Collection

dingty-zz opened the pull request:
We have implemented the following modes for GC
1. GC_TYPE_OFF
2. GC_TYPE_VACUUM, which runs a background vacuum thread to clean garbage
3. GC_TYPE_COOPERATIVE, in which every thread does GC
4. GC_TYPE_EPOCH, which associates each txn with an epoch and manages garbage per epoch

The followings are the files/functions that are relevant to this code review:
1. src/backend/gc/gc_manager.h/cpp
2. src/backend/gc/gc_manager_factory.h/cpp
3. src/backend/common/epoch.h/cpp
4. src/backend/concurrency/optimistic_txn_manager.h
        4.1 OptimisticTxnManager::BeginTransaction
        4.2  OptimisticTxnManager::EndTransaction
5. src/backend/concurrency/transaction_manager.h
         5.1 TransactionManager::RecycleTupleSlot

Test cases:
tests/garbage_collection/ 9 test cases, three per GC mode

Note: The tests written pass when run as standalone. When run using make check, which runs them with valgrind, the tests (along with a couple of the other tests) fail. For all the failed test (including the tests we have written and the already existing ones), valgrind throws an error from the Boost lock-free queue. This may be an error in the Boost lock-free queue or in our code (and the code which the couple of other tests check, which we haven't written). For now, we have not been able to isolate the cause.

Commit:
refator IN+NestLoop

Commit:
refore rm peloton_client.h

Commit:
test protobuf sending queryplan

Commit:
test protobuf

Commit:
test protobuf with nested message

Commit:
test protobuf simple

Commit:
test protobuf 3 level nested

Commit:
add serialization and de.. for class

Commit:
add postgres structure for protobuf

Commit:
function for postgres structure

Commit:
parse postgres structure

Commit:
parse and create postgres structure. It's tough work

Commit:
checkpoint

Commit:
Make the ValueType part of AbstractExpression's constructor. This means that type information for the return value of ALL expressions will be known before query evaluation. This seems like an obvious thing, but was missing. It'll also greatly help when doing codegen

Commit:
add framework for sending tuple back

Commit:
add test

Commit:
add test

Commit:
merge with master is done

Commit:
merge with master is done

Commit:
add query.proto for compile

Commit:
add queryplan.proto for compile

Commit:
remove IN for nestedloop and change the new model for tupledesc

Commit:
add new and delete for unit test

Commit:
add include memory for travisci check

Commit:
add include memory for travisci check

Commit:
remove malloc

Commit:
add materialization for LogicalTile

Commit:
execute query is done

Commit:
return tile

Commit:
remove deleted file and code

Commit:
merge with master

Commit:
merge with master is done

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
remove IN+subquery for index scan

dingty committed:
GetMemoryFootprint is done, but needs testing.
Didn't do a merge mvcc.

saurabhkadekodi committed:
Added naive co-operative GC

saurabhkadekodi committed:
merge

saurabhkadekodi committed:
resolved conflicts

saurabhkadekodi committed:
Used GCManagerFactory to fetch type

saurabhkadekodi committed:
Untested implementation of epoch based GC

saurabhkadekodi committed:
fixed smalled epoch cleaned to be cid instead of oid

Commit:
Added SetGCType and called it with config param

saurabhkadekodi committed:
Fixed bugs in epoch GC

fixed mapping of epoch_map_ to CID

Commit:
Fixed vacuum GC

saurabhkadekodi committed:
Changed cuckoo hashmap to boost flat_map

Commit:
Take GC type from config file

Commit:
Merge branch 'master' of github.com:saurabhkadekodi/peloton-gc

Commit:
GetMemoryFootprint raw code done

dingty committed:
Compile code

dingty committed:
Add memoryfp call sites

saurabhkadekodi committed:
Fixed epoch hang bug

Fixed flat_map usage in optmistic_transaction_manager

saurabhkadekodi committed:
merge

saurabhkadekodi committed:
added epoch as another GC option

changed default state of GC to off

Commit:
fix merge problem

Commit:
fix merge problem

Commit:
Small fix

Commit:
Oops

Commit:
Merge pull request #130 from Michael-Tieying-Zhang/master

Sending query plan between peloton nodes

dingty committed:
GMF causes TATP to crash

saurabhkadekodi committed:
Moved GetMemoryFoot[rint to GCStop

omegaga committed:
change Makefile to be aware of proto file changes

Commit:
Merge pull request #159 from omegaga/fixbug

change Makefile to be aware of proto file changes

larryxiao committed:
update travis to compile log message #161

apavlo committed:
A bunch of fixes for LOG_TRACE arguments that were not compiling. #161

larryxiao committed:
fix string format #161

larryxiao committed:
Update .travis.yml

revert changes because too many logging messages in travis

apavlo committed:
Merge pull request #162 from larryxiao/master

update travis to compile all log message #161

Commit:
Small bug fixes

GetMemorySize bug fix

Commit:
Removed networking stuff from peloton dml

Commit:
Added integration testing script

Commit:
Updated script

Commit:
Updated integration script

Commit:
Update test_peloton.sh

Commit:
Update .travis.yml

Commit:
Update .travis.yml

Commit:
Update .travis.yml

Commit:
Update .travis.yml

Commit:
Removed erroneous break from PerformGC

dingty committed:
Add one test case

dingty committed:
Merge branch 'new_master' of github.com:saurabhkadekodi/peloton-gc into new_master

dingty committed:
3 test cases in total

dingty committed:
Refactor code and add more test cases

dingty committed:
Fix coop max_cid and test cases

dingty committed:
Refactor test cases

dingty committed:
Merge branch 'new_master' of github.com:saurabhkadekodi/peloton-gc into new_master

Commit:
Commments

Commit:
More commenting

Commit:
Merge branch 'new_master' of github.com:saurabhkadekodi/peloton-gc into new_master

dingty committed:
Initialize gc_manager

dingty committed:
Merge branch 'new_master' of github.com:saurabhkadekodi/peloton-gc into new_master

Commit:
More comments

dingty committed:
Refactor

dingty committed:
One more comment

dingty committed:
Format code

yudun commented:
@dingty @saurabhkadekodi @rajatkateja Could you please resolve all the conflicts and make a pr again? We are not able to merge your code with yingjun's concurrency control, which is used in my version and I don't know how to resolve them.

rajatkateja commented:
I don't think its required to merge the code. You can do the code review even without merging. That's how I was doing the review for your code as well. Can you explain why merging is a prerequisite?

yudun commented:
Because we would like to check if the code works.

yudun-zz commented on src/backend/common/epoch.cpp:
```diff
@@ -0,0 +1,123 @@
+        return retval;
+      } else {
+        return retval;
+      }
```
Line 91-117 is essentially the same as line 57-86, except for that you use max_cid rather than largest_epoch here. So it would be better if you refract them and create a function, say PerformGC(....) to remove duplicated code.

yudun-zz commented on src/backend/gc/gc_manager.cpp:
```diff
@@ -14,129 +14,190 @@
+  // Loop infinitely and perform GC
+  while (1) {
+    PerformGC();
+    std::this_thread::sleep_for(std::chrono::seconds(5));
```
Please replace the magic number 5 with a macro

yudun-zz commented on src/backend/gc/gc_manager.cpp:
```diff
@@ -14,129 +14,190 @@
     return;
   }
-  
+  // Populate the tuple metadata structure
```
Probably using a constructor instead is better?

yudun commented:
BTW, it seems that you only records and reclaims unused ItemPointer for your GC, right? But as we know that data like varchar resides in varlenpool, did you implement any GC mechenism for this part of data?

pmenon committed:
Fix cast expression

pmenon committed:
Cleanup

Commit:
Merge pull request #168 from pmenon/expression-refactor

Expression refactor

yudun commented:
General Comments for your tests:
1. I find that you have comprehensive test for your three modes of GC, which is good. But I also find that many duplicated code exist in your 9 files, which should definitely be improved --- you can create a file named gc_tests_utils to put these functions like InsertTuple, DeleteTuple and UpdateTuple, and you don't have to copy and paste them in all the files.

2. If you create the gc_tests_utils file, you don't have to separate your 9 tests in 9 files. This test should be a single module and it would be better that you put your 9 tests in  9 TEST_F function in a single testing file.

yudun-zz commented on tests/garbage_collection/gc_delete_test_coop.cpp:
```diff
@@ -0,0 +1,169 @@
+  auto *table = ExecutorTestsUtil::CreateTable(1024);
+  auto &manager = catalog::Manager::GetInstance();
+  storage::Database db(DEFAULT_DB_ID);
+  manager.AddDatabase(&db);
```
If you use this way to create a new db in your test, probably you should also add `manager.DropDatabaseWithOid(current_db_oid);` 
In the end of this tests.

We meet similar problem that you meet during our development: a single test success, but when run together with others, it fails. Later we find that it is because we didn't clean the data of temporary created database clearly, and after add this line in the end of our test, it works perfectly well. You can refer to line 892 in our tests/catalog/constraints_test.cpp and have a try. Hope this will help you

yudun commented:
Code Review Summary
---
** General Questions **

Does the code work?
Don't know -- conflicts for this pull request

Is all the code easily understood?
Yes, for most of them.

Is there any redundant or duplicate code?
Yes, as illustrated in above comments

Is the code as modular as possible?
Yes.

Can any global variables be replaced?
No.

Is there any commented out code?
No.

Is it using proper debug log functions?
Yes.

** Documentation Questions **

Do comments describe the intent of the code?
Yes.

Are all functions commented?
Yes.

Is any unusual behavior described?
Specified in the pr comment.

Is the use of 3rd-party libraries documented?
Yes.

Is there any incomplete code?
Yes. GCManager::DeleteTupleFromIndexes

** Testing Questions **

Do tests exist and are they comprehensive?
Yes.

Are the tests actually testing the feature?
Yes.

Are they relying on hardcoded answers?
Yes, hardcoded answers exist in some places in the testing code.

What is the code coverage?
Garbage Collection

saurabhkadekodi committed:
merge with CMU-DB master

saurabhkadekodi committed:
switched from boost lockfree queue to moodycamel lockfree queue

added moodycamel lockfree queue to third_party library

Commit:
Refactored Code according to code reviews

Commit:
Fixed optimistic txn which broke the gc tests

Commit:
Fixed bug in GC

Commit:
Added combined test files

apavlo commented:
@yudun Do you guys have any specific comments for the code? You don't have to actually run it.

yudun commented:
Hi Andy, @apavlo , all the comments we gave was above the code summary, please have a check

Commit:
Removed partial GC from some txn mangers

Now only optimistic txn manager has a (complete) implementation of GC.
No other txn manager has GC implementations

Commit:
Reverted the combined tests

Commit:
Removed the combined test files
